### DISCUSSION TITLE
Modifying Certain Keys for ZMK Implementation

### DISCUSSION THREAD

**Discussion by robert-hill** [manna-harbour/miryoku#224] [Q&A] [answered]
I have what I hope is a question for this group, if not please send me the right direction!

I use Miryoku on my 5 col Corne build and am finally feeling comfortable using it day to day, but there are two little modifications I would like to do to really make it perfect and I just can figure it out at all. I use a QWERTY, flipped layer build with inverted-T nav, and I really want to map Escape to the left outer thumb key, Tab to the middle left hand thumb key, and then move space to the inner right hand thumb key and backspace to the middle right hand thumb key. I swear I have read the docs and several posts by the developer of the layout on how to do this but it has so far defeated me. 

The main reasoning for this is to keep the same muscle memory that I have for my Kinesis Advantage keyboards which is my main daily driver. The Corne is my mobile/travel KB and when I end up using it I find myself sending messages prematurely at work which is massively embarrassing, and then when I get back to my Advantages I have the opposite problem and it can take some time to get back in the groove again.

Any help would be appreciated here, or if there is some sort of dumbed down tutorial or something to follow that would be cool too. I am a software dev so I can handle most of this on my own but this is just confusing the heck out of me and I don't quite understand some of choices for how it is intended that ZMK implementation to be modified.

**Comment by manna-harbour** [comment, marked as answer]
There are 2 issues: customising layers, and how to best deal with compatibility with your KA. I'll address customising layers first.

To customise layers, see https://github.com/manna-harbour/miryoku_zmk#customisation, in particular the _Substitute Layer_ section. The steps are as follows:

- Find the definition of the layer to be modified in https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/miryoku_babel/miryoku_layer_alternatives.h. In the case of Base QWERTY flip, that would be the definition of `MIRYOKU_ALTERNATIVES_BASE_QWERTY_FLIP`, at https://github.com/manna-harbour/miryoku_zmk/blob/d0c749b53d90d44b05a9ac442a690125b3a0f009/miryoku/miryoku_babel/miryoku_layer_alternatives.h#L57-L61. In general you can work it out by name, or by working through the `#if`s in https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/miryoku_babel/miryoku_layer_selection.h, in your case https://github.com/manna-harbour/miryoku_zmk/blob/d0c749b53d90d44b05a9ac442a690125b3a0f009/miryoku/miryoku_babel/miryoku_layer_selection.h#L9 for the Base layer, https://github.com/manna-harbour/miryoku_zmk/blob/d0c749b53d90d44b05a9ac442a690125b3a0f009/miryoku/miryoku_babel/miryoku_layer_selection.h#L10 for flip, https://github.com/manna-harbour/miryoku_zmk/blob/d0c749b53d90d44b05a9ac442a690125b3a0f009/miryoku/miryoku_babel/miryoku_layer_selection.h#L27 for QWERTY, leading to https://github.com/manna-harbour/miryoku_zmk/blob/d0c749b53d90d44b05a9ac442a690125b3a0f009/miryoku/miryoku_babel/miryoku_layer_selection.h#L28.
- Copy the definition to https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/custom_config.h. The contents would be
```
#define MIRYOKU_ALTERNATIVES_BASE_QWERTY_FLIP \
&kp Q,             &kp W,             &kp E,             &kp R,             &kp T,             &kp Y,             &kp U,             &kp I,             &kp O,             &kp P,             \
U_MT(LGUI, A),     U_MT(LALT, S),     U_MT(LCTRL, D),    U_MT(LSHFT, F),    &kp G,             &kp H,             U_MT(LSHFT, J),    U_MT(LCTRL, K),    U_MT(LALT, L),     U_MT(LGUI, SQT),   \
U_LT(U_BUTTON, Z), U_MT(RALT, X),     &kp C,             &kp V,             &kp B,             &kp N,             &kp M,             &kp COMMA,         U_MT(RALT, DOT),   U_LT(U_BUTTON, SLASH),\
U_NP,              U_NP,              U_LT(U_FUN, DEL),  U_LT(U_NUM, BSPC), U_LT(U_SYM, RET),  U_LT(U_MOUSE, TAB),U_LT(U_NAV, SPACE),U_LT(U_MEDIA, ESC),U_NP,              U_NP
```
- Change `MIRYOKU_ALTERNATIVES_BASE_QWERTY_FLIP` to `MIRYOKU_LAYER_BASE`. This will now override the `MIRYOKU_ALPHAS` option and use this custom definition instead. The contents would be
```
#define MIRYOKU_LAYER_BASE \
&kp Q,             &kp W,             &kp E,             &kp R,             &kp T,             &kp Y,             &kp U,             &kp I,             &kp O,             &kp P,             \
U_MT(LGUI, A),     U_MT(LALT, S),     U_MT(LCTRL, D),    U_MT(LSHFT, F),    &kp G,             &kp H,             U_MT(LSHFT, J),    U_MT(LCTRL, K),    U_MT(LALT, L),     U_MT(LGUI, SQT),   \
U_LT(U_BUTTON, Z), U_MT(RALT, X),     &kp C,             &kp V,             &kp B,             &kp N,             &kp M,             &kp COMMA,         U_MT(RALT, DOT),   U_LT(U_BUTTON, SLASH),\
U_NP,              U_NP,              U_LT(U_FUN, DEL),  U_LT(U_NUM, BSPC), U_LT(U_SYM, RET),  U_LT(U_MOUSE, TAB),U_LT(U_NAV, SPACE),U_LT(U_MEDIA, ESC),U_NP,              U_NP
```
- Rearrange the keycodes as desired. E.g. to swap Tap and Esc, swap the locations of `TAB` and `ESC` in the definition. The contents would be
```
#define MIRYOKU_LAYER_BASE \
&kp Q,             &kp W,             &kp E,             &kp R,             &kp T,             &kp Y,             &kp U,             &kp I,             &kp O,             &kp P,             \
U_MT(LGUI, A),     U_MT(LALT, S),     U_MT(LCTRL, D),    U_MT(LSHFT, F),    &kp G,             &kp H,             U_MT(LSHFT, J),    U_MT(LCTRL, K),    U_MT(LALT, L),     U_MT(LGUI, SQT),   \
U_LT(U_BUTTON, Z), U_MT(RALT, X),     &kp C,             &kp V,             &kp B,             &kp N,             &kp M,             &kp COMMA,         U_MT(RALT, DOT),   U_LT(U_BUTTON, SLASH),\
U_NP,              U_NP,              U_LT(U_FUN, DEL),  U_LT(U_NUM, BSPC), U_LT(U_SYM, RET),  U_LT(U_MOUSE, ESC),U_LT(U_NAV, SPACE),U_LT(U_MEDIA, TAB),U_NP,              U_NP
```

**Reply by robert-hill** [reply]
Thank you so much for the incredibly in depth answer. I am going to give all this a try today. As for using the same keymap on my 360Pro, I may try that in the future.

**Comment by manna-harbour** [comment]
For compatibility with your KA, it's best to use the same keymap everywhere. Miryoku ZMK is supported on adv360pro, and support for `kinesis/stapelberg` can be added to Miryoku QMK. A stock KA can be used with Miryoku KMonad or via https://github.com/manna-harbour/xmk.

**Comment by ryangco** [comment]
Hi. I would like to ask what I'm doing wrong trying to change layout and a layer.
First, I want to change corne layout and add mapping to the unused keys. This works fine on its own.
Then I want to change some of the rgb controls to laptop brightness inc dec. I followed the guide using the substitute layer section of the documentation and added the code to the bottom but it does not work. 
What am I doing wrong? Thanks. 

```
// Copyright 2021 Manna Harbour
// https://github.com/manna-harbour/miryoku

#if !defined (MIRYOKU_LAYOUTMAPPING_CORNE)

#define XXX &none

#define MIRYOKU_LAYOUTMAPPING_CORNE( \
     K00, K01, K02, K03, K04,      K05, K06, K07, K08, K09, \
     K10, K11, K12, K13, K14,      K15, K16, K17, K18, K19, \
     K20, K21, K22, K23, K24,      K25, K26, K27, K28, K29, \
     N30, N31, K32, K33, K34,      K35, K36, K37, N38, N39 \
) \
&kp MINUS  K00  K01  K02  K03  K04       K05  K06  K07  K08  K09  &kp PIPE \
&kp ESC  K10  K11  K12  K13  K14       K15  K16  K17  K18  K19  &kp EQUAL \
&kp PLUS  K20  K21  K22  K23  K24       K25  K26  K27  K28  K29  &kp UNDER \
               K32  K33  K34       K35  K36  K37

#define MIRYOKU_KLUDGE_TOPROWCOMBOS_LEFTPINKIE 2 3
#define MIRYOKU_KLUDGE_TOPROWCOMBOS_LEFTINNERINDEX 3 4
#define MIRYOKU_KLUDGE_TOPROWCOMBOS_RIGHTINNERINDEX 7 8
#define MIRYOKU_KLUDGE_TOPROWCOMBOS_RIGHTPINKIE 8 9

#define MIRYOKU_KLUDGE_BOTTOMROWCOMBOS_LEFTPINKIE 26 27
#define MIRYOKU_KLUDGE_BOTTOMROWCOMBOS_LEFTINNERINDEX 27 28
#define MIRYOKU_KLUDGE_BOTTOMROWCOMBOS_RIGHTINNERINDEX 31 32
#define MIRYOKU_KLUDGE_BOTTOMROWCOMBOS_RIGHTPINKIE 32 33

#define MIRYOKU_KLUDGE_THUMBCOMBOS_LEFT 37 38
#define MIRYOKU_KLUDGE_THUMBCOMBOS_RIGHT 39 40


#endif

#define MIRYOKU_CLIPBOARD_WIN
#define MIRYOKU_EXTRA_QWERTY

#define MIRYOKU_LAYER_MEDIA \
U_BOOT,            &u_to_U_TAP,       &u_to_U_EXTRA,     &u_to_U_BASE,      U_NA,              U_RGB_TOG,         &kp C_BRIGHTNESS_DEC,         U_RGB_HUI,         U_RGB_SAI,         &kp C_BRIGHTNESS_INC,         \
&kp LGUI,          &kp LALT,          &kp LCTRL,         &kp LSHFT,         U_NA,              U_EP_TOG,          &kp C_PREV,        &kp C_VOL_DN,      &kp C_VOL_UP,      &kp C_NEXT,        \
U_NA,              &kp RALT,          &u_to_U_FUN,       &u_to_U_MEDIA,     U_NA,              &u_out_tog,        &u_bt_sel_0,       &u_bt_sel_1,       &u_bt_sel_2,       &u_bt_sel_3,       \
U_NP,              U_NP,              U_NA,              U_NA,              U_NA,              &kp C_STOP,        &kp C_PP,          &kp C_MUTE,        U_NP,              U_NP`
```

**Reply by ryangco** [reply]
Moved the #define MIRYOKU_LAYER_MEDIA \ to the top of the custom_config and it worked. Not sure if that's all I did wrong.

**Reply by manna-harbour** [reply]
It was probably a lack of line terminating newline on the last line.
